File: scala/aoc/src/main/scala/aoc2015/day06.scala
package aoc2015

import scala.collection.mutable.HashSet
import scala.collection.mutable.HashMap

sealed trait CommandType
case object TurnOn extends CommandType
case object TurnOff extends CommandType
case object Toggle extends CommandType
case class Point(x: Int, y: Int)
case class Command(t: CommandType, from: Point, to: Point)

def part1(commands: Array[Command]): Long =
  val set = HashSet[Point]()
  for 
    cmd <- commands
    y <- cmd.from.y to cmd.to.y
    x <- cmd.from.x to cmd.to.x
  do
    val p = Point(x, y)
    cmd.t match
      case TurnOn => set += p
      case TurnOff => set -= p
      case Toggle => if set.contains(p) then set -= p else set += p
  set.size

def part2(commands: Array[Command]): Long =
  val vs = HashMap[Point, Long]()
  for
    cmd <- commands
    y <- cmd.from.y to cmd.to.y
    x <- cmd.from.x to cmd.to.x
  do
    val p = Point(x, y)
    cmd.t match
      case TurnOn if vs.contains(p) => vs(p) += 1L
      case TurnOn => vs(p) = 1L
      case TurnOff if vs.contains(p) => if vs(p) <= 1 then vs -= p else vs(p) -= 1L
      case TurnOff => // Skip
      case Toggle if vs.contains(p) => vs(p) += 2L
      case Toggle => vs(p) = 2L
  vs.values.sum

@main def main =
  val turnon_re = "^turn on (\\d+),(\\d+) through (\\d+),(\\d+)$".r
  val turnoff_re = "^turn off (\\d+),(\\d+) through (\\d+),(\\d+)$".r
  val toggle_re = "^toggle (\\d+),(\\d+) through (\\d+),(\\d+)$".r
  val commands = io.Source.fromFile("input.txt").getLines()
    .map(_.trim()).filter(_.nonEmpty).map:
      case turnon_re(x1, y1, x2, y2) => Command(TurnOn, Point(x1.toInt, y1.toInt), Point(x2.toInt, y2.toInt))
      case turnoff_re(x1, y1, x2, y2) => Command(TurnOff, Point(x1.toInt, y1.toInt), Point(x2.toInt, y2.toInt))
      case toggle_re(x1, y1, x2, y2) => Command(Toggle, Point(x1.toInt, y1.toInt), Point(x2.toInt, y2.toInt))
    .toArray
  
  println(s"2015 Day 6, Part 1: ${part1(commands)}")
  println(s"2015 Day 6, Part 2: ${part2(commands)}")

File: rust/aoc/src/aoc2015/day06.rs
use regex::Regex;
use crate::utils::profile;

const BOUND: usize = 1000;

#[derive(Debug)]
pub enum Command {
  TurnOn((usize, usize), (usize, usize)),
  TurnOff((usize, usize), (usize, usize)),
  Toggle((usize, usize), (usize, usize)),
}
use Command::*;
use crate::common::day::Day;

pub struct Grid<T: Default + Copy> {
  grid: Vec<Vec<T>>,
}

impl<T: Default + Copy> Grid<T> {
  fn new() -> Self {
    Grid {
      grid: vec![vec![T::default(); BOUND]; BOUND],
    }
  }

  fn eval(&mut self, from: (usize, usize), to: (usize, usize), f: fn(T) -> T) {
    let (x1, y1) = from;
    let (x2, y2) = to;
    for x in x1..=x2 {
      for y in y1..=y2 {
        self.grid[y][x] = f(self.grid[y][x])
      }
    }
  }

  fn each<F>(&self, f: &mut F)
    where
      F: FnMut(T),
  {
    for x in 0..BOUND {
      for y in 0..BOUND {
        f(self.grid[y][x]);
      }
    }
  }

  fn reset(&mut self) {
    for x in 0..BOUND {
      for y in 0..BOUND {
        self.grid[y][x] = T::default();
      }
    }
  }
}


struct Day06;

impl Day for Day06 {
  type Input = Vec<Command>;
  type Output = u64;
  fn day() -> u32 { 6 }
  fn year() -> u32 { 2015 }
  
  fn parse_input(input: String) -> Vec<Command> {
    let turn_on_re = Regex::new(r"^turn\s*on\s*(\d+),(\d+)\s*through\s*(\d+),(\d+)$").unwrap();
    let turn_off_re = Regex::new(r"^turn\s*off\s*(\d+),(\d+)\s*through\s*(\d+),(\d+)$").unwrap();
    let toggle_re = Regex::new(r"^toggle\s*(\d+),(\d+)\s*through\s*(\d+),(\d+)$").unwrap();
    input.trim()
      .split("\n")
      .map(|line| line.trim())
      .filter(|line| !line.is_empty())
      .map(|line| {
        if let Some(captures) = turn_on_re.captures(line) {
          let x1 = captures.get(1).unwrap().as_str().parse::<usize>().unwrap();
          let y1 = captures.get(2).unwrap().as_str().parse::<usize>().unwrap();
          let x2 = captures.get(3).unwrap().as_str().parse::<usize>().unwrap();
          let y2 = captures.get(4).unwrap().as_str().parse::<usize>().unwrap();
          TurnOn((x1, y1), (x2, y2))
        } else if let Some(captures) = turn_off_re.captures(line) {
          let x1 = captures.get(1).unwrap().as_str().parse::<usize>().unwrap();
          let y1 = captures.get(2).unwrap().as_str().parse::<usize>().unwrap();
          let x2 = captures.get(3).unwrap().as_str().parse::<usize>().unwrap();
          let y2 = captures.get(4).unwrap().as_str().parse::<usize>().unwrap();
          TurnOff((x1, y1), (x2, y2))
        } else if let Some(captures) = toggle_re.captures(line) {
          let x1 = captures.get(1).unwrap().as_str().parse::<usize>().unwrap();
          let y1 = captures.get(2).unwrap().as_str().parse::<usize>().unwrap();
          let x2 = captures.get(3).unwrap().as_str().parse::<usize>().unwrap();
          let y2 = captures.get(4).unwrap().as_str().parse::<usize>().unwrap();
          Toggle((x1, y1), (x2, y2))
        } else {
          panic!("Unmatched command pattern: {line}");
        }
      }).collect::<Vec<Command>>()
  }
  
  fn part1(commands: &Vec<Command>) -> u64 {
    let mut grid: Grid<bool> = Grid::<bool>::new();
    for c in commands {
      match c {
        &TurnOn(from, to) => grid.eval(from, to, |_| true),
        &TurnOff(from, to) => grid.eval(from, to, |_| false),
        &Toggle(from, to) => grid.eval(from, to, |b| !b),
      }
    }

    let mut res = 0u64;
    grid.each(&mut |v| {
      if v { res += 1; }
    });
    res
  }

  fn part2(commands: &Vec<Command>) -> u64 {
    let mut grid: Grid<u32> = Grid::<u32>::new();
    for c in commands {
      match c {
        &TurnOn(from, to) => grid.eval(from, to, |v| v + 1),
        &TurnOff(from, to) => grid.eval(from, to, |v| if v == 0 { 0 } else { v - 1 }),
        &Toggle(from, to) => grid.eval(from, to, |v| v + 2),
      }
    }

    let mut res = 0u64;
    grid.each(&mut |v| {
      res += v as u64;
    });
    res
  }
}

pub mod tests {
  use super::*;

  #[test]
  fn test_part1() {}

  #[test]
  fn run() {
    Day06::run();
  }
}
